plutus-contract/src/Language/Plutus/Contract/Effects/WatchAddress.hs

```diff
@@ -31,32 +36,98 @@ import           Language.Plutus.Contract.Util              (loopM)
 type AddressSymbol = "address"
 
 type HasWatchAddress s =
-    ( HasType AddressSymbol (Address, Tx) (Input s)
-    , HasType AddressSymbol Address (Output s)
+    ( HasType AddressSymbol AddressChangeResponse (Input s)
+    , HasType AddressSymbol AddressChangeRequest (Output s)
     , ContractRow s)
 
-type WatchAddress = AddressSymbol .== ((Address, Tx), Address)
+type WatchAddress = AddressSymbol .== (AddressChangeResponse, AddressChangeRequest)
+
+-- | Information about transactions that spend or produce an output at
+--   an address. The transactions are returned in the order they appear
+--   in the ledger (oldest first).
+data AddressChangeResponse =
+    AddressChangeResponse
+        { acrAddress :: Address
+        , acrSlot    :: Slot
+        , acrTx      :: Tx
+        }
+        deriving stock (Eq, Generic, Show)
+        deriving anyclass (ToJSON, FromJSON)
+
+instance Pretty AddressChangeResponse where
+    pretty AddressChangeResponse{acrAddress, acrTx, acrSlot} =
+        hang 2 $ vsep
+            [ "Address:" <+> pretty acrAddress
+            , "Slot:" <+> pretty acrSlot
+            , "TxId:" <+> pretty (txId acrTx)
+            ]
+
+data AddressChangeRequest =
+    AddressChangeRequest
+        { acreqSlot    :: Slot
+        , acreqAddress :: Address
+        , acreqTxId    :: Maybe TxId
+        }
+        deriving stock (Eq, Generic, Show, Ord)
+        deriving anyclass (ToJSON, FromJSON)
+
+instance Pretty AddressChangeRequest where
+    pretty AddressChangeRequest{acreqSlot, acreqAddress, acreqTxId} =
+        hang 2 $ vsep
+            [ "Slot:" <+> pretty acreqSlot
+            , "Address:" <+> pretty acreqAddress
+            , "TxId:" <+> pretty acreqTxId
+            ]
+
+{-| Wait for the next transaction that changes an address. The meaning of
+    "next transaction" for an 'AddressChangeRequest' depends on the value
+    of 'acreqTxId'.
+
+    1. If 'acreqTxId' is 'Just txid' then @b@ is the next transaction if
+    @b@ spends or produces an output at the given address and @b@ is the first
+    transaction to be added to the ledger in or after 'acreqSlot' and
+    after 'txid'.
+    2. If 'acreqTxId' is 'Nothing' then @b@ is the next transaction if
+    b spends or produces an output at the given address and b is the first
+    transaction to be added to the ledger in or after 'acreqSlot'.
+
+    Some implications of this definition are:
+
+    * If the transaction 'txid' is never added to the ledger then (1) will
+    never return
+    * If the slot 'acreqSlot' contains more than one transaction at the address
+    then only (1) will ever return transactions other than the first one.
```

Comment:
Perhaps dumb question: why not return all the transactions that modify the address in the first *slot* after the specified one? They can still be ordered, so you can just take the first one.

Reply:
That way we can answer the request before the slot is over and don't have to wait. So if two PABs run the same contract, for different roles, then they can have more than one interaction per slot.

I think in the medium term we will need to deal with block numbers as well as slot numbers, and maybe then we can always wait until the block is over, and don't need the `txId` anymore as you said below.

Reply:
In Shelley there's at most one block per slot, so doing interactions "inside" a slot doesn't make much sense: either you've seen the block or you haven't. So I don't think what you said is possible anyway, and we might as well just give you everything per slot. Indeed, then we can simplify and just say it gives you all the transactions that touch that address in the slot that you specify, which you might have to wait for if it's in the future.

Reply:
That might actually be wrong, checking now. And even if there are multiple blocks per slot, you won't *know* about any of them until the slot is over. So you can't see the first block, then respond and do something in that same slot.

Reply:
Ah right, that would make things a little easier on our side.

Reply:
I deleted the txID argument, we can still change from `Slot` to `Block` later on if we need to.